src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -1,43 +1,219 @@
+        RequestDelegate next,
+        IGraphQLTextSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        GraphQLHttpMiddlewareOptions options,
+        IEnumerable<IWebSocketHandler<TSchema>>? webSocketHandlers = null)
+        : base(next, serializer, options, webSocketHandlers)
+    {
+        _documentExecuter = documentExecuter ?? throw new ArgumentNullException(nameof(documentExecuter));
+        _serviceScopeFactory = serviceScopeFactory ?? throw new ArgumentNullException(nameof(serviceScopeFactory));
+        var getRule = new HttpGetValidationRule();
+        _getValidationRules = DocumentValidator.CoreRules.Append(getRule).ToArray();
+        _getCachedDocumentValidationRules = new[] { getRule };
+        var postRule = new HttpPostValidationRule();
+        _postValidationRules = DocumentValidator.CoreRules.Append(postRule).ToArray();
+        _postCachedDocumentValidationRules = new[] { postRule };
+    }
 
+    private static IEnumerable<IWebSocketHandler<TSchema>> CreateWebSocketHandlers(
+        IGraphQLSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        IServiceProvider provider,
+        IHostApplicationLifetime hostApplicationLifetime,
+        GraphQLHttpMiddlewareOptions options)
+    {
+        if (hostApplicationLifetime == null)
+            throw new ArgumentNullException(nameof(hostApplicationLifetime));
+        if (provider == null)
+            throw new ArgumentNullException(nameof(provider));
+
+        return new IWebSocketHandler<TSchema>[] {
+            new WebSocketHandler<TSchema>(serializer, documentExecuter, serviceScopeFactory, options,
+                hostApplicationLifetime, provider.GetService<IWebSocketAuthenticationService>()),
+        };
+    }
+
+    ******************************/
+
+    /// <inheritdoc/>
+    protected override async Task<ExecutionResult> ExecuteScopedRequestAsync(HttpContext context, GraphQLRequest request, IDictionary<string, object?> userContext)
+    {
+        using var scope = _serviceScopeFactory.CreateScope();
+        return await ExecuteRequestAsync(context, request, scope.ServiceProvider, userContext);
+    }
+
+    /// <inheritdoc/>
+    protected override async Task<ExecutionResult> ExecuteRequestAsync(HttpContext context, GraphQLRequest request, IServiceProvider serviceProvider, IDictionary<string, object?> userContext)
+    {
+        if (request?.Query == null)
+        {
+            return new ExecutionResult
+            {
+                Errors = new ExecutionErrors {
+                    new QueryMissingError()
+                }
+            };
+        }
+        var opts = new ExecutionOptions
+        {
+            Query = request.Query,
+            Variables = request.Variables,
+            Extensions = request.Extensions,
+            CancellationToken = context.RequestAborted,
```

Comment:
The reason (to me) to have `GetCancellationToken` which you deleted is to pass custom token into all components including GraphQL's ones just from beginning of http request. Also it applies to
` return _serializer.WriteAsync(context.Response.Body, result, context.RequestAborted);` 
that was changed from
 `return _serializer.WriteAsync(context.Response.Body, result, GetCancellationToken(context));`

I watched through MS code and noticed that `HttpContext.RequestAborted` is settable property so I actually can change my code to set it instead of creating new token and use it side-by-side with one from `HttpContext.RequestAborted`

```csharp
 httpContext.RequestAborted = timeout.HasValue // some custom timeout taken from wherever
                    ? CancellationTokenSource.CreateLinkedTokenSource(httpContext.RequestAborted, new CancellationTokenSource(timeout.Value).Token).Token
                    : httpContext.RequestAborted;
```

And I know that now I can modify execution options on each request like

```csharp
builder.ConfigureExecutionOptions(opt => opt.CancellationToken = ..../*can pull some data from service provider*/)
```

What do you think about directly modifying `HttpContext.RequestAborted` property?

Reply:
To clarify - my goal now to understand if I can remove my custom middleware that inherits from `GraphQLHttpMiddleware`. I migrated all my "old" code to `ConfigureExecutionOptions` (by the way, this led to additional memory allocations, but it's not so scary). Only two pieces left - handling token and log execution results:
```csharp
 public class GraphQLHttpInterceptionMiddleware<TSchema> : GraphQLHttpMiddleware<TSchema>
        where TSchema : ISchema
    {
        private readonly IExecutionResultInterceptor[] _interceptors;
        private readonly IServiceCallContextAccessor _serviceCallContextAccessor;

        public GraphQLHttpInterceptionMiddleware(
            IGraphQLTextSerializer serializer,
            IEnumerable<IExecutionResultInterceptor> executionResultInterceptors,
            IServiceCallContextAccessor serviceCallContextAccessor)
            : base(serializer)
        {
            _interceptors = executionResultInterceptors.ToArray();
            _serviceCallContextAccessor = serviceCallContextAccessor;
        }

        protected override Task RequestExecutedAsync(in GraphQLRequestExecutionResult requestExecutionResult)
        {
            if (_interceptors.Length > 0)
            {
                var req = new Request(requestExecutionResult.Request.Query, requestExecutionResult.Request.OperationName, requestExecutionResult.Request.Variables);
                foreach (var interceptor in _interceptors)
                    interceptor.InterceptExecutionResult(new RequestExecutionResult(req, requestExecutionResult.Result, requestExecutionResult.Elapsed, requestExecutionResult.IndexInBatch));
            }

            return base.RequestExecutedAsync(requestExecutionResult);
        }

        protected override CancellationToken GetCancellationToken(HttpContext context)
            => _serviceCallContextAccessor.ServiceCallContext?.CancellationToken ?? CancellationToken.None;
    }
```

Actually I have only one `IExecutionResultInterceptor` implementation that logs GraphQL request/response and elapsed. What could you suggest to write the same feature to log request/response and elapsed?

Reply:
WOW, well I can just switch my middleware from abstract one calling arbitrary set of interceptors to one that only logs requests and elapsed! Like `GraphQLHttpMiddlewareWithLogs` does, that is all 🤦

Reply:
> What do you think about directly modifying `HttpContext.RequestAborted` property?

Seems to make sense to me.  You could probably write a piece of ASP.NET Core middleware to change it:

```cs
public Task InvokeAsync(HttpContext context)
{
    // or whatever
    using var cts = new CancellationTokenSource();
    var token = cts.CancellationToken;

    var oldToken = context.RequestAborted;
    context.RequestAborted = token;
    try
    {
        await next(context);
    }
    catch (OperationCancelledException) when (token.IsCancellationRequired)
    {
    }
    context.RequestAborted = oldToken;
}
```

This has a benefit of affecting all ASP.Net Core middleware, not just GraphQL, and the token is disposed properly.

Personally I would pay attention to ensuring that a thrown OperationCancelledException when the original RequestAborted is unchanged is caught or not, as demonstrated above.

Reply:
You are actually talking about how to distinguish server cancellation (some server-side timeout or whatever withing http pipeline) from client cancellation (aborted http connection). Right?

Reply:
Right, because the ASP.NET Core pipeline only catches `OperationCancelledException`s when it is the client (aborted http connection).  For good practices, I don't like unhandled exceptions to propagate up the indefinitely.  Since ASP.NET Core wouldn't handle the "service-side timeout" exception (caused by `token.ThrowIfCancellationRequested()` deeper in the chain), it should be handled by the code where the cancellation token started.